base/broadcast.jl

```diff
@@ -612,63 +776,183 @@ julia> string.(("one","two","three","four"), ": ", 1:4)
+    copy(instantiate(_broadcast_incremental(bc)))
+end
+function _broadcast_incremental(bc::Broadcasted)
+    not_nested(bc) && return broadcast(bc.f, Tuple(bc.args)...)
+    Broadcasted(bc.f, mapTupleLL(_broadcast_incremental, bc.args))
+end
+_broadcast_incremental(x) = x
+
+# When ElType is not concrete, use narrowing. Use the first output
+# value to determine the starting output eltype; copyto_nonleaf!
+# will widen `dest` as needed to accommodate later values.
+function copy_nonleaf(bc::Broadcasted{Style,ElType}) where {Style,ElType}
+    iter = CartesianIndices(axes(bc))
+    state = start(iter)
+    if done(iter, state)
+        # if empty, take the ElType at face value
+        return broadcast_similar(Style(), ElType, axes(bc), bc)
+    end
+    # Initialize using the first value
+    I, state = next(iter, state)
+    val = _broadcast_getindex(bc, I)
+    dest = broadcast_similar(Style(), typeof(val), axes(bc), bc)
+    dest[I] = val
+    # Now handle the remaining values
+    copyto_nonleaf!(dest, bc, iter, state, 1)
 end
 
-# When ElType is not concrete, use narrowing. Use the first element of each input to determine
-# the starting output eltype; the _broadcast! method will widen `dest` as needed to
-# accommodate later values.
-function broadcast_nonleaf(f, s::NonleafHandlingTypes, ::Type{ElType}, shape::Indices, As...) where ElType
-    nargs = length(As)
-    iter = CartesianIndices(shape)
-    if isempty(iter)
-        return Base.similar(Array{ElType}, shape)
+## general `copyto!` methods
+# The most general method falls back to a method that replaces Style->Nothing
+# This permits specialization on typeof(dest) without introducing ambiguities
+@inline copyto!(dest::AbstractArray, bc::Broadcasted) =
+    copyto!(dest, convert(Broadcasted{Nothing}, bc))
+
+# Performance optimization for the Scalar case
+@inline function copyto!(dest::AbstractArray, bc::Broadcasted{<:Union{Scalar,Unknown},ElType,Nothing,Nothing}) where ElType
+    if not_nested(bc)
+        if bc.f === identity && bc.args isa TupleLL{<:Any,TupleLLEnd} # only a single input argument to broadcast!
+            # broadcast!(identity, dest, val) is equivalent to fill!(dest, val)
+            return fill!(dest, bc.args.head)
+        else
+            args = Tuple(bc.args)
+            @inbounds for I in eachindex(dest)
+                dest[I] = bc.f(args...)
+            end
+            return dest
+        end
     end
-    keeps, Idefaults = map_newindexer(shape, As)
-    st = start(iter)
-    I, st = next(iter, st)
-    val = f([ _broadcast_getindex(As[i], newindex(I, keeps[i], Idefaults[i])) for i=1:nargs ]...)
-    if val isa Bool
-        dest = Base.similar(BitArray, shape)
-    else
-        dest = Base.similar(Array{typeof(val)}, shape)
+    # Fall back to the default implementation
+    copyto!(dest, instantiate(instantiate_axes(bc)))
+end
+
+# Specialize this method if all you want to do is specialize on typeof(dest)
```

Comment:
Is this comment really true?  I think you'd just want to specialize `copyto!(::MyArray, ::Broadcast)`, not `::Broadcast{Nothing}`.  Or am I missing something?